apps/src/code-studio/pd/form_components/FormComponent.jsx

```diff
@@ -231,11 +258,21 @@ export default class FormComponent extends React.Component {
   static getDynamicallyRequiredFields(data) {
     return [];
   }
+
+  /**
+   * Override in derived classes
+   * @param {Object} data - form data for this page
+   * @returns {Object} - custom error messages per field
+   */
+  static getErrorMessages(data) {
+    return {};
+  }
```

Comment:
CC @Hamms I added the ability to supply custom error messages and fail validation (i.e. for bad format). See screenshots and example usage above for phone number and zip.